server/src/main/java/org/elasticsearch/search/SearchService.java

```diff
@@ -1097,16 +1112,22 @@ private DefaultSearchContext createSearchContext(
         return searchContext;
     }
 
-    static boolean supportsParallelCollection(ResultsType resultsType, SearchSourceBuilder source) {
+    int determineMaximumNumberOfSlices(ExecutorService executor, ShardSearchRequest request, ResultsType resultsType) {
+        return executor instanceof ThreadPoolExecutor tpe
+            && supportsParallelCollection(resultsType, request.source(), this.enableQueryPhaseParallelCollection)
+                ? tpe.getMaximumPoolSize()
+                : 1;
+    }
```

Comment:
By extracting this logic into a function, I was able to introduce a boundary to allow testing via `testEnableQueryPhaseParallelCollection()`.

Reply:
One comment on this, out of paranoia: I recently added a test here to verify that supportsParallelCollection affects the max number of slices. I wanted to verify that despite returning the right value, the method had indeed an effect on the number of slices generated. I wonder if that test needs updating now that we have an additional method. I am tempted to suggest to fold supportsparallelCollection into your new method and unify the tests as well. What do you think?

Reply:
> I recently added a test here to verify that supportsParallelCollection affects the max number of slices.
Can you link to the code you are referring to?

Personally, I think the function `supportsParallelCollection` is not well named, as it's unclear what the subject is and this invites confusion. (e.g. "what supports parallel collection? The instance?)

However, I do like having small functions that do one thing, as they are easier to test and reason about, and their name can describe the logic contained therin. 

What do you think about just renaming `supportsParallelCollection` to something like `isParallelCollectionSupportedForResults`?  That way, it's more clear what's going on, and it's harder to misuse in future.